File: backend/src/main/java/com/electricity/billing/service/BillService.java
package com.electricity.billing.service;

import com.electricity.billing.dto.request.AddBillRequest;
import com.electricity.billing.dto.response.BillResponse;
import com.electricity.billing.dto.response.BillSelectionSummaryResponse;
import com.electricity.billing.dto.response.BulkUploadResultResponse;
import com.electricity.billing.dto.response.PageResponse;
import com.electricity.billing.entity.enums.BillStatus;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDate;
import java.util.List;

public interface BillService {

    BillResponse addBill(AddBillRequest request);

    BulkUploadResultResponse bulkUpload(MultipartFile file);

    PageResponse<BillResponse> myBills(BillStatus status, int page, int size);

    PageResponse<BillResponse> myBillHistory(LocalDate fromDate, LocalDate toDate, BillStatus status,
                                              String sortBy, int page, int size);

    BillSelectionSummaryResponse getSelectionSummary(List<Long> billIds);

    PageResponse<BillResponse> adminSearch(String consumerNumber, String customerCode, BillStatus status,
                                            int page, int size);

    byte[] exportBillHistory(String consumerNumber, String format);

    BillResponse getBill(Long id);
}

File: frontend/src/app/core/services/bill.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AddBillRequest, Bill, BillSelectionSummary, BulkUploadResult } from '../models/bill.model';
import { PageResponse } from '../models/page-response.model';

@Injectable({ providedIn: 'root' })
export class BillService {
  private readonly baseUrl = '/api/bills';

  constructor(private readonly http: HttpClient) {}

  addBill(request: AddBillRequest): Observable<Bill> {
    return this.http.post<Bill>(this.baseUrl, request);
  }

  bulkUpload(file: File): Observable<BulkUploadResult> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<BulkUploadResult>(`${this.baseUrl}/bulk-upload`, formData);
  }

  myBills(status: string | null, page: number, size: number): Observable<PageResponse<Bill>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (status) params = params.set('status', status);
    return this.http.get<PageResponse<Bill>>(`${this.baseUrl}/me`, { params });
  }

  myBillHistory(fromDate: string | null, toDate: string | null, status: string | null, sortBy: string,
                page: number, size: number): Observable<PageResponse<Bill>> {
    let params = new HttpParams().set('sortBy', sortBy).set('page', page).set('size', size);
    if (fromDate) params = params.set('fromDate', fromDate);
    if (toDate) params = params.set('toDate', toDate);
    if (status) params = params.set('status', status);
    return this.http.get<PageResponse<Bill>>(`${this.baseUrl}/me/history`, { params });
  }

  selectionSummary(billIds: number[]): Observable<BillSelectionSummary> {
    return this.http.post<BillSelectionSummary>(`${this.baseUrl}/me/selection-summary`, billIds);
  }

  adminSearch(consumerNumber: string, customerCode: string, status: string | null,
              page: number, size: number): Observable<PageResponse<Bill>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (consumerNumber) params = params.set('consumerNumber', consumerNumber);
    if (customerCode) params = params.set('customerCode', customerCode);
    if (status) params = params.set('status', status);
    return this.http.get<PageResponse<Bill>>(this.baseUrl, { params });
  }

  exportBillHistory(consumerNumber: string, format: 'csv' | 'pdf'): Observable<Blob> {
    const params = new HttpParams().set('consumerNumber', consumerNumber).set('format', format);
    return this.http.get(`${this.baseUrl}/export`, { params, responseType: 'blob' });
  }

  getById(id: number): Observable<Bill> {
    return this.http.get<Bill>(`${this.baseUrl}/${id}`);
  }
}
